### PR TITLE
rgw: beast frontend can listen on multiple endpoints

### PR DESCRIPTION
**Pull request by cbodley**
extends the frontend configuration to store options in a multimap, so that `port` and `endpoint` can be specified multiple times. the beast frontend creates a separate `Listener` for each port/endpoint given, and this `Listener` instance is passed as an extra argument when `accept()`ing new connections. the civetweb frontend also supports multiple `port` options

Fixes: http://tracker.ceph.com/issues/22779

### CODE REVIEW

File: src/rgw/rgw_asio_frontend.cc [hunk 1]
```diff
@@ -220,64 +226,104 @@ class AsioFrontend {
...
+  auto port = std::strtoul(input, &end, 10);
+  if (port > std::numeric_limits<unsigned short>::max()) {
+    ec.assign(ERANGE, boost::system::system_category());
+  } else if (port == 0 && end == input) {
+    ec.assign(EINVAL, boost::system::system_category());
+  }
+  return port;
+}
+
+tcp::endpoint parse_endpoint(std::experimental::string_view input,
```
Comment (rzarzynski) [inline src/rgw/rgw_asio_frontend.cc]: Does Ceph have any policy on `std::experimental`? What about using `boost::string_view` instead?

Comment (cbodley) [reply src/rgw/rgw_asio_frontend.cc]: ah, thanks for reminding me about this part

i had found that `boost::asio::ip::make_address()` wouldn't accept `boost::string_view` or `std::string_view`, so this was the only form that would compile. but now that i look closer (http://www.boost.org/doc/libs/1_66_0/boost/asio/detail/string_view.hpp), there are typedefs for `string_view`/`basic_string_view` in namespace `boost::asio` that i can use instead

**Comment by cbodley** [comment]
thanks @rzarzynski, updated to use the same `boost::asio::string_view` typedef used by `boost::asio::ip::make_address()`

**Comment by cbodley** [comment]
passed qa: http://pulpito.ceph.com/cbodley-2018-01-31_10:32:24-rgw-wip-cbodley-testing-distro-basic-smithi/ with environment failures rerun http://pulpito.ceph.com/cbodley-2018-02-05_09:43:36-rgw-wip-cbodley-testing-distro-basic-smithi/

**Pull request merged**